server.js

```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+const express = require('express');
+const bodyParser = require('body-parser');
+const Auth = require('./auth.js');
+
+const app = express();
+app.use(bodyParser.json());
```

Comment:
Welcome to your first middleware 🎉 .

Middleware is basically stuff that happens from the request hitting the app to the request actually hitting your own code.

This usually entails serializing objects, routing, logging... But you can `use` _:D_ it for whatever you want. I, for instance, like to use it for decoupling validations and auth checks.

Here, as you can see, it's just serializing your request to JSON.

Let's do our first one.

# Responsibilites

It looks like my favorite word. You're probably sick of me saying it.

You're already putting too much stuff here:
- Instantiates the server
- Puts the middleware
- Runs the server
- Instantiates `Auth`
- Does Routing
- Has controllers
- Does views

And it's only 29 lines long and just has one endpoint. This is not going to scale.

## Extracting the Routes

Create a `routes` folder. It seems like you're namespacing this one to `user` (you should use the plural for objects, so I'll go for `users` from now on).

Perfect, let's create a namespaced routing system for users:
```js
// routes/users.js
const express = require('express');
const router = express.Router();
const Auth = require('../auth');

const auth = new Auth();

// Notice I don't need the `/users` namespace
app.post('/register', async (req, res) => {
  const credentials = req.body;
  if (credentials.username && credentials.password) {
    const result = await auth.registerUser(credentials.username, credentials.password);
    if (result.success) {
      res.status(201).json({status: 201, user: credentials.username});
    } else {
      res.status(409).json({status: 409, error: result.message});
    }
  } else {
    res.status(400).json({status: 400, error: 'Data is invalid'});
  }
});
```

Now add your routing as middleware in `app.js`:
```js
// app.js
const express = require('express');
const bodyParser = require('body-parser');
const usersRouter = require('./routes/users');

const app = express();
app.use(bodyParser.json());
app.use('/users', usersRouter);  // <-- :D

const server = app.listen(process.env.PORT || 8080, () => {
  const port = server.address().port;
  console.log(`App now running on port ${port}`);
});
```

Ok, a lot less responsibilities for `app.js`. Now it only has to set the middleware and run the server.

But our Routing now has to do controller and view stuff. That's not cool...

Reply:
_The answer to that will come in the next episode_

Reply:
Ok, next episode is _ON_.

# Controllers

In today's episode we're going to start creating _Controllers_. As you may know, controllers orchestrate different parts of the application together, with one caveat: 
_controllers are not necessarily supposed to know what they're doing_.

I see this a lot, even in our application. People think that because controllers are the ones in charge of orchestrating stuff they should know the nitty-gritty details of everything. That results in a controller being 50 lines long —when it should be 5-10 lines long—.

Controllers are like managers. They tell people what to do, but don't necessarily know how to do it, they just know who to ask and how to orchestrate them. Tim knows it has to tell _you_ to change the color of a button, it doesn't mean he _knows_ how to do it (or even should).

In a well written application, those _responsibilities_ (my favorite word) go to _Service Objects_.

There's a lot of ways to do SOs, but I like them to have just one responsibility (or just one reason to change), following the _Single Responsibility Principle_ (from now on **SRP**). 
I also like them to have the same _interface_ (following the _Liskov Substitution Principle_) which would be having just one public method that it's always the same (in Ruby we use `.call`, because it's used heavily by other things in Ruby).
This also means you're not forcing your _clients_ to rely on methods they _don't need_ (thus following the _Interface Segregation Principle_).
If you add things like what you did with the `db` object, you would be using SOs to _invert the dependency_ and rely on abstractions instead of low level details (following the _Dependency Inversion Principle_) then you can basically see we're aiming for a _[SOLID](https://en.wikipedia.org/wiki/SOLID)_ design :D.

Okay, let's refactor. We have the following right now:

```js
// routes/users.js
router.post('/register', async (req, res) => {
  const credentials = req.body;
  if (credentials.username && credentials.password) {
    const result = await auth.registerUser(credentials.username, credentials.password);
    if (result.success) {
      res.status(201).json({status: 201, user: credentials.username});
    } else {
      res.status(409).json({status: 409, error: result.message});
    }
  } else {
    res.status(400).json({status: 400, error: 'Data is invalid'});
  }
});
```

A routing module should not care about all of this, let's extract this into a controller:
```js
// controllers/users.js
const Auth = require('../auth');
const auth = new Auth();

exports.register = async (req, res) => {
  const credentials = req.body;

  if (credentials.username && credentials.password) {
    const result = await auth.registerUser(
      credentials.username,
      credentials.password
    );
    if (result.success) {
      res.status(201).json({ status: 201, user: credentials.username });
    } else {
      res.status(409).json({ status: 409, error: result.message });
    }
  } else {
    res.status(400).json({ status: 400, error: 'Data is invalid' });
  }
};
```
And now our router looks like this:
```js
// routes/users.js
const express = require('express');
const router = express.Router();
const usersController = require('../controllers/users');

router.post('/register', usersController.register);

module.exports = router;
```

_HOT DAMN_

Now the router only cares about routing. My parents were right: I'm always finding ways to avoid responsibilities.

But now I look at the controller and it still feels wrong. I should be able to look at it and just _know_ what it does, but I get lost in a sea of `if` statements... 

This controller knows _too much_... We'll have to deal with it 🔫

But that will be on the next episode:
We'll delve on Service Objects, Models (even _top_ models), some more middleware and we'll probably break your _Auth_ object apart :D

Reply:
app and the users route looks so good now! This entire section was fantastic! Tons of good reading and I see what you mean about the users controller I have now, I'm excited to keep refactoring these down.